iwyu_test_util.py

```diff
@@ -419,6 +436,112 @@ def _GetLaunchArguments(cc_file):
   return shlex.split(args)
 
 
+def _CheckForCompileTimeMacro(macro):
+    """This assumes that the local clang++ and the include-what-you-use
+    under test are using the same libclang, and therefore the same default libraries.
+
+    Look for the specified macro in the clang output.
+    """
+    txt = '#include <cstddef>'
+
+    result = subprocess.run(['clang', '-dM', '-E', '-x', 'c++', '-'], capture_output=True,
+                            input=txt, text=True)
+
+    return " "+macro+" " in str(result.stdout)
+
+
+def _CheckExtraCommandLine(args):
+  """Check if particular options are on the command line"""
+  env_iwyu_extra_args = os.getenv('IWYU_EXTRA_ARGS')
+  iwyu_extra_args = shlex.split(env_iwyu_extra_args)
+  standardised_args = ' '.join(iwyu_extra_args)
+  return args in standardised_args
+
+
+def _EvalCondition(cond):
+  """Evaluate the input condition.
+
+  All conditions comprise a label and a single argument. The supported
+  tests are:
+
+  IwyuHasTargetSupportFor arg - whether include-what-you-use supports
+                                the architecture 'arg' as a target.
+
+  TestingCXXSTL (libc++|libstdcpp|canonical) - whether the current
+                                test is compiling against a particular
+                                C++ standard library.
+  """
+  # Some helper functions to handle conditions
+  def isTestingLibCxx():
+    return _CheckExtraCommandLine('-stdlib=libc++') or _CheckForCompileTimeMacro('_LIBCPP_VERSION')
+  def isTestingLibstdcpp():
+    return _CheckExtraCommandLine('-stdlib=libstdc++') or _CheckForCompileTimeMacro('__GLIBCXX__')
+  def isTestingCanonicalClangMappings():
+    return _CheckExtraCommandLine('-Xiwyu --experimental=clang_mappings')
+  def iwyuHasTargetSupportFor(tgt):
+    iwyu = _GetIwyuPath()
+    result = subprocess.run([iwyu, '--print-targets'], capture_output=True)
+    return tgt in str(result.stdout)
+
+  m = _IWYU_COND.match(cond)
+  if m:
+    label = m.group(1)
+    arg = m.group(2)
+
+    if label == 'IwyuHasTargetSupportFor':
+      return iwyuHasTargetSupportFor(arg)
+    elif label == 'TestingCXXSTL':
+      if arg == 'libcxx':
+        return isTestingLibCxx()
+      elif arg == 'libstdcpp':
+        return isTestingLibstdcpp()
+      elif arg == 'canonical':
+        return isTestingCanonicalClangMappings()
+
+  raise SyntaxError("unsupported condition \"%s\"" % cond)
+
+
+def CheckSkipConditions(cc_file):
+  """Check the test skip conditions for a source file from its contents.
+
+  Returns None if the test should be run. If the test should be
+  skipped, returns the condition as a string that the caller can print.
+  """
+  with open(cc_file) as it:
+    found_block = False
+    # Find the next '// IWYU_*: ' line.
+    for lineno, line in enumerate(it):
+      m = _IWYU_TEST_SKIP_RE.match(line)
+      if m:
+        found_block = True
+        typ = m.group(1)
+        cond = m.group(2)
+
+        try:
```

Comment:
I'd prefer to separate the parsing from the evaluation here. It looks like this function currently only evaluates the first skip-condition, but in theory there could be any number of them, right?

Reply:
Sorry, I conflated two concerns:

- Please split this up so there's a function that only parses and returns skip directives (maybe as a tuple `(skip_type, skip_cond)`)
- When evaluating the directives, we should probably evaluate all of them and skip if any hold

Reply:
FYI, this evaluated multiple conditions, but exited on the first condition indicating we should skip. Splitting as requested.